[WIP] UI Improvements

PSI-Rockin opened the pull request:
This PR adds a settings window that allows users to provide system files, such as boot ROMs and NAND images, through the UI. It also pops up a message box whenever a fatal error occurs and stops emulation, rather than crashing the whole program. The emulator can be booted with or without a 3DS cartridge through the menubar. Finally, emulation now runs on a separate thread so that the UI remains responsive even when the emulator lags.

There remains some work to be done. It would be nice to have an FPS counter, for instance, and some of the code needs cleanup. Furthermore, there is the question of what to do with command-line arguments, as they are currently commented out. The best option is likely to just save arguments into the settings.

PSI-Rockin committed:
Qt: Add settings window for boot ROMs, NAND images, and SD images

Comment out all of the emulator junk for now

PSI-Rockin committed:
Qt: Move Emulator object to separate thread

The emulator can be initialized through settings now, but it currently does not run yet.

PSI-Rockin committed:
Qt: Execute emulation on separate thread

This is pretty barebones, but at least it's functional

PSI-Rockin committed:
Emulator: Close mounted images when re-opening them

Some cleanup for EmuWindow

PSI-Rockin committed:
Qt: Pass per-frame settings to Emuthread, such as input

Fix bugs causing SD path to not be saved properly and settings window to be shown when pressing up

profi200 commented:
Yeah, just override the settings with command line args. Another option is splitting it into a stripped down version and the Qt GUI version similar to what citra does.

PSI-Rockin committed:
CMake: Attempt to fix build

PSI-Rockin committed:
Qt: Use QCommandLineParser to parse arguments

PSI-Rockin committed:
I2C: Implement power button interrupt

Also fix some command line bugs

PSI-Rockin committed:
I2C: Implement HOME button

PSI-Rockin committed:
EmuThread: Fix hang when MCU reboot signal is sent

PSI-Rockin committed:
ARM: Simplify and speed up debugging logic

Prevent crashes on O3DS NANDs by using heuristics to determine where the pid and CodeSet pointer are in KProcess

PSI-Rockin committed:
Qt: Implement FPS counter

PSI-Rockin committed:
Emulator: Stub PDN registers used by O3DS during power-off
